Improve `sbtype` and `oplmode` descriptions & clean up `sblaster.cpp`

johnnovak opened the pull request:
# Description

This is some heavy-duty cleanup of `sblaster.cpp` because it looked like the dog's breakfast and I couldn't stand it anymore. The biggest non-superficial improvements are probably the strictly typed enums and making all function args const where possible.

Hopefully I'll be a little bit happier the next time I'm touching this file.

The `sbtype` and `oplmode` descriptions are also vastly improved.

No functional changes.


# Manual testing

Tried a few games and demos, no crashes and the Sound Blaster emulation still works.


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

japsmits commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1189,1332 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
+		ctmixer_update_volumes();
+		break;
+
+	case 0x06: // FM output selection
+	           // Somewhat obsolete with dual OPL SBpro + FM volume (SB2 Only)
+		// volume controls both channels
```
Correct comment formatting.

johnnovak replied:
This was clang-format being a bit confused, actually 😄

japsmits commented on src/hardware/tandy_sound.cpp:
```diff
@@ -303,22 +303,30 @@ void TandyDAC::ChangeMode()
-		if (const auto sample_rate = tandy_psg_clock_hz / regs.clock_divider;
-		    sample_rate < dac_max_sample_rate_hz) {
+
+		if (const auto new_sample_rate = tandy_psg_clock_hz / regs.clock_divider;
+		    new_sample_rate < dac_max_sample_rate_hz) {
```
Should we log something if the new_sample_rate is out-of-bounds?

johnnovak replied:
Yeah, I might put an assert there, ok. But I did not want to do deeper refactoring this time around.

johnnovak replied:
Actually, we don't want to log out-of-bounds values. See the comment a few lines above this line.

I remember a discussion about this, this would result in hundreds of log entries in some games that set very high sample rates first, which then "stabilises" to something within the normal range.

> 	// Typical sample rates are 1.7. 5.5, 11, and rarely 22 KHz. Although
	// several games (one being OutRun) set instantaneous rates above
	// 100,000 Hz, we throw these out as they can cause garbage high
	// frequency harmonics and also cause problems for the Speex Resampler.
	// For example, a clock divider value of 8 (which is valid) produces a
	// 450 KHz sampling rate, which is way beyond what Speex can handle.
	//

japsmits commented on src/hardware/tandy_sound.cpp:
```diff
@@ -303,22 +303,30 @@ void TandyDAC::ChangeMode()
+
 					channel->Enable(true);
-					// LOG_MSG("TANDYDAC: playback started with freqency %f, volume %f", sample_rate, vol);
+					// LOG_MSG("TANDYDAC: playback started
+					// with freqency %f, volume %f",
+					// sample_rate, vol);
```
This won't compile if uncommented.

japsmits reviewed (commented):
Nice one! Comments are mostly issues created by reformatting, with the occasional small code-change.

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -806,236 +871,301 @@ static void PlayDMATransfer(uint32_t bytes_requested)
 			if (samples & 1) {
-				// Carry over the dangling sample into the next round, or
+				// Carry over the dangling sample into the next
+				// round, or
```
Dunno. Fell off the truck 🤷🏻 😆

johnnovak replied:
Sorted it out, the comment continues in the else branch, but it was misleading without the `...`.

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1189,1332 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
+	case DmaMode::Pcm16BitAliased: sb.dma.mul = (1 << SbShift) * 2; break;
 	default:
-		LOG(LOG_SB,LOG_ERROR)("DSP:Illegal transfer mode %d", mode);
+		LOG(LOG_SB, LOG_ERROR)("DSP:Illegal transfer mode %d", mode);
```
Not now, I ran out of steam 😅 These log statements don't normally show up in our logs; these must be specially enabled for "dev logging". 

Job for someone else on another day...

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1189,1332 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
-			LOG(LOG_SB,LOG_NORMAL)("DSP Unhandled SB16ASP command %X (set mode register to %X)",sb.dsp.cmd,sb.dsp.in.data[0]);
+		if (sb.type == SBType::SB16) {
+			// SB16 ASP set mode register
+			if ((sb.dsp.in.data[0] & 0xf1) == 0xf1) {
+				asp_init_in_progress = true;
+			} else {
+				asp_init_in_progress = false;
+			}
```
Cool, I'll do this, but did not really want to do deeper refactoring this time, just getting rid of the superficial stuff first.

johnnovak replied:
Actually, maybe next time 😄 This is big enough already.

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -806,236 +871,301 @@ static void PlayDMATransfer(uint32_t bytes_requested)
-	        "bytes_to_read=%u, bytes_read = %u, samples = %u, frames = %u, dma.left = %u",
-	        CardType(), sb.dma.mode, sb.dma.stereo, sb.dma.sign, bytes_requested,
-	        bytes_to_read, bytes_read, samples, frames, sb.dma.left);
+	        "bytes_to_read=%u, bytes_read = %u, samples = %u, frames = %u,
+	dma.left = %u", log_prefix(), sb.dma.mode, sb.dma.stereo, sb.dma.sign,
+	bytes_requested, bytes_to_read, bytes_read, samples, frames,
+	sb.dma.left);
```
Unfortunately, clang-format will ignore `#if 0 / #endif` blocks and I couldn't find an option to change that. Anyway, using this is better than commenting out log statements.

johnnovak commented:
Okay @japsmits I've added one last commit after the one that addresses your PR comments.

I wasn't aware of 46ce6be073 which simplifies the handling of "falsy" and "truthy" setting values. So there's no need to explicitly handle the "none" case with `parse_bool_setting` in my `sbtype` and `oplmode` parsing function which simplifies things considerably.

After some consideration, I like this behaviour because it simplifies the parsing and makes the handling of "falsy"and "truthy" values consistent—one just needs to be aware of this.

japsmits commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1199,1339 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
-		if (	((sb.type == SBT_PRO1 || sb.type == SBT_PRO2) && sb.mixer.index==0x0c) || /* Input control on SBPro */
-			(sb.type == SBT_16 && sb.mixer.index >= 0x3b && sb.mixer.index <= 0x47)) /* New SB16 registers */
+		if (((sb.type == SBType::SBPro1 || sb.type == SBType::SBPro2) &&
+		     sb.mixer.index == 0x0c) || // Input control on SBPro */
```
```suggestion
		     sb.mixer.index == 0x0c) || // Input control on SBPro
```

japsmits commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1199,1339 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
+		if (((sb.type == SBType::SBPro1 || sb.type == SBType::SBPro2) &&
+		     sb.mixer.index == 0x0c) || // Input control on SBPro */
+		    (sb.type == SBType::SB16 && sb.mixer.index >= 0x3b &&
+		     sb.mixer.index <= 0x47)) { // New SB16 registers */
```
```suggestion
		     sb.mixer.index <= 0x47)) { // New SB16 registers
```

japsmits commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1199,1339 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
+
+	case 0x15:
+		// Wari hack. Waru uses this one instead of 0x14, but some
+		// weird stuff going on there anyway */
```
```suggestion
		// weird stuff going on there anyway
```

japsmits commented on src/hardware/sblaster.cpp:
```diff
@@ -1059,938 +1199,1339 @@ static void set_channel_rate_hz(const uint32_t requested_rate_hz)
-		              sb.freq));
+
+	case 0x91:
+		// Singe Cycle 8-Bit DMA High speed DAC */
+		// Note: 0x91 is documented only for DSP ver.2.x and 3.x,
+		// not 4.x */
```
```suggestion
		// Singe Cycle 8-Bit DMA High speed DAC
		// Note: 0x91 is documented only for DSP ver.2.x and 3.x,
		// not 4.x
```

japsmits reviewed (approved):
It's just some comments. 😸

johnnovak commented:
Thanks for the thorough review and the suggestions @japsmits 🎉
